

<!DOCTYPE html>


<html lang="en" >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

    <title>Analysing the brain cells in light stimulated mice &#8212; scPCA 0.3.3 documentation</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "light";
  </script>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="../_static/styles/theme.css?digest=927b94d3fcb96560df09" rel="stylesheet" />
<link href="../_static/styles/bootstrap.css?digest=927b94d3fcb96560df09" rel="stylesheet" />
<link href="../_static/styles/pydata-sphinx-theme.css?digest=927b94d3fcb96560df09" rel="stylesheet" />

  
  <link href="../_static/vendor/fontawesome/6.1.2/css/all.min.css?digest=927b94d3fcb96560df09" rel="stylesheet" />
  <link rel="preload" as="font" type="font/woff2" crossorigin href="../_static/vendor/fontawesome/6.1.2/webfonts/fa-solid-900.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../_static/vendor/fontawesome/6.1.2/webfonts/fa-brands-400.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../_static/vendor/fontawesome/6.1.2/webfonts/fa-regular-400.woff2" />

    <link rel="stylesheet" type="text/css" href="../_static/pygments.css" />
    <link rel="stylesheet" href="../_static/styles/sphinx-book-theme.css?digest=14f4ca6b54d191a8c7657f6c759bf11a5fb86285" type="text/css" />
  
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="../_static/scripts/bootstrap.js?digest=927b94d3fcb96560df09" />
<link rel="preload" as="script" href="../_static/scripts/pydata-sphinx-theme.js?digest=927b94d3fcb96560df09" />
  <script src="../_static/vendor/fontawesome/6.1.2/js/all.min.js?digest=927b94d3fcb96560df09"></script>

    <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js"></script>
    <script src="../_static/jquery.js"></script>
    <script src="../_static/underscore.js"></script>
    <script src="../_static/_sphinx_javascript_frameworks_compat.js"></script>
    <script src="../_static/doctools.js"></script>
    <script src="../_static/sphinx_highlight.js"></script>
    <script src="../_static/scripts/sphinx-book-theme.js?digest=5a5c038af52cf7bc1a1ec88eea08e6366ee68824"></script>
    <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script>window.MathJax = {"tex": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true}, "options": {"ignoreHtmlClass": "tex2jax_ignore|mathjax_ignore|document", "processHtmlClass": "tex2jax_process|mathjax_process|math|output_area"}}</script>
    <script defer="defer" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    <script>DOCUMENTATION_OPTIONS.pagename = 'notebooks/angelidis';</script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Analysing the lung cell populations in aging mice" href="hrvatin.html" />
    <link rel="prev" title="Integrating the human pancreas dataset" href="pancreas.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <a class="skip-link" href="#main-content">Skip to main content</a>
  
  <div id="pst-scroll-pixel-helper"></div>

  
  <button type="button" class="btn rounded-pill" id="pst-back-to-top">
    <i class="fa-solid fa-arrow-up"></i>
    Back to top
  </button>

  
  <input type="checkbox"
          class="sidebar-toggle"
          name="__primary"
          id="__primary"/>
  <label class="overlay overlay-primary" for="__primary"></label>
  
  <input type="checkbox"
          class="sidebar-toggle"
          name="__secondary"
          id="__secondary"/>
  <label class="overlay overlay-secondary" for="__secondary"></label>
  
  <div class="search-button__wrapper">
    <div class="search-button__overlay"></div>
    <div class="search-button__search-container">
<form class="bd-search d-flex align-items-center"
      action="../search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         id="search-input"
         placeholder="Search..."
         aria-label="Search..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form></div>
  </div>
  
    <nav class="bd-header navbar navbar-expand-lg bd-navbar">
    </nav>
  
  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      <div class="bd-sidebar-primary bd-sidebar">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
    
    
  </div>
  
    <div class="sidebar-primary-items__start sidebar-primary__section">
        <div class="sidebar-primary-item">
  

<a class="navbar-brand logo" href="../index.html">
  
  
  
  
  
    <p class="title logo__title">scPCA 0.3.3 documentation</p>
  
</a></div>
        <div class="sidebar-primary-item"><nav class="bd-links" id="bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item navbar-nav active">
        <p aria-level="2" class="caption" role="heading"><span class="caption-text">API</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../models.html">Models</a></li>
<li class="toctree-l1"><a class="reference internal" href="../tools.html">Tools</a></li>
<li class="toctree-l1"><a class="reference internal" href="../plots.html">Plotting</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Tutorials</span></p>
<ul class="current nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="kang.html">Analysing interferon stimulated PBMCs using scPCA</a></li>
<li class="toctree-l1"><a class="reference internal" href="pancreas.html">Integrating the human pancreas dataset</a></li>
<li class="toctree-l1 current active"><a class="current reference internal" href="#">Analysing the brain cells in light stimulated mice</a></li>
<li class="toctree-l1"><a class="reference internal" href="hrvatin.html">Analysing the lung cell populations in aging mice</a></li>
</ul>

    </div>
</nav></div>
    </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
  </div>
  
  <div id="rtd-footer-container"></div>


      </div>
      
      <main id="main-content" class="bd-main">
        
        

<div class="sbt-scroll-pixel-helper"></div>

          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item"><label class="sidebar-toggle primary-toggle btn btn-sm" for="__primary" title="Toggle primary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
  <span class="fa-solid fa-bars"></span>
</label></div>
      
    </div>
  
  
    <div class="header-article-items__end">
      
        <div class="header-article-item">

<div class="article-header-buttons">





<div class="dropdown dropdown-download-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Download this page">
    <i class="fas fa-download"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="../_sources/notebooks/angelidis.ipynb" target="_blank"
   class="btn btn-sm btn-download-source-button dropdown-item"
   title="Download source file"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file"></i>
  </span>
<span class="btn__text-container">.ipynb</span>
</a>
</li>
      
      
      
      
      <li>
<button onclick="window.print()"
  class="btn btn-sm btn-download-pdf-button dropdown-item"
  title="Print to PDF"
  data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file-pdf"></i>
  </span>
<span class="btn__text-container">.pdf</span>
</button>
</li>
      
  </ul>
</div>




<button onclick="toggleFullScreen()"
  class="btn btn-sm btn-fullscreen-button"
  title="Fullscreen mode"
  data-bs-placement="bottom" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>


<script>
document.write(`
  <button class="btn btn-sm navbar-btn theme-switch-button" title="light/dark" aria-label="light/dark" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="theme-switch nav-link" data-mode="light"><i class="fa-solid fa-sun fa-lg"></i></span>
    <span class="theme-switch nav-link" data-mode="dark"><i class="fa-solid fa-moon fa-lg"></i></span>
    <span class="theme-switch nav-link" data-mode="auto"><i class="fa-solid fa-circle-half-stroke fa-lg"></i></span>
  </button>
`);
</script>

<script>
document.write(`
  <button class="btn btn-sm navbar-btn search-button search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass fa-lg"></i>
  </button>
`);
</script>
<label class="sidebar-toggle secondary-toggle btn btn-sm" for="__secondary"title="Toggle secondary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="fa-solid fa-list"></span>
</label>
</div></div>
      
    </div>
  
</div>
</div>
              
              

<div id="jb-print-docs-body" class="onlyprint">
    <h1>Analysing the brain cells in light stimulated mice</h1>
    <!-- Table of contents -->
    <div id="print-main-content">
        <div id="jb-print-toc">
            
            <div>
                <h2> Contents </h2>
            </div>
            <nav aria-label="Page">
                <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#Data-preprocessing">Data preprocessing</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#Factor-analysis">Factor analysis</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#Factor-0">Factor 0</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#Factor-1">Factor 1</a></li>
</ul>
</li>
</ul>
            </nav>
        </div>
    </div>
</div>

              
                
<div id="searchbox"></div>
                <article class="bd-article" role="main">
                  
  
<style>
/* CSS for nbsphinx extension */

/* remove conflicting styling from Sphinx themes */
div.nbinput.container div.prompt *,
div.nboutput.container div.prompt *,
div.nbinput.container div.input_area pre,
div.nboutput.container div.output_area pre,
div.nbinput.container div.input_area .highlight,
div.nboutput.container div.output_area .highlight {
    border: none;
    padding: 0;
    margin: 0;
    box-shadow: none;
}

div.nbinput.container > div[class*=highlight],
div.nboutput.container > div[class*=highlight] {
    margin: 0;
}

div.nbinput.container div.prompt *,
div.nboutput.container div.prompt * {
    background: none;
}

div.nboutput.container div.output_area .highlight,
div.nboutput.container div.output_area pre {
    background: unset;
}

div.nboutput.container div.output_area div.highlight {
    color: unset;  /* override Pygments text color */
}

/* avoid gaps between output lines */
div.nboutput.container div[class*=highlight] pre {
    line-height: normal;
}

/* input/output containers */
div.nbinput.container,
div.nboutput.container {
    display: -webkit-flex;
    display: flex;
    align-items: flex-start;
    margin: 0;
    width: 100%;
}
@media (max-width: 540px) {
    div.nbinput.container,
    div.nboutput.container {
        flex-direction: column;
    }
}

/* input container */
div.nbinput.container {
    padding-top: 5px;
}

/* last container */
div.nblast.container {
    padding-bottom: 5px;
}

/* input prompt */
div.nbinput.container div.prompt pre {
    color: #307FC1;
}

/* output prompt */
div.nboutput.container div.prompt pre {
    color: #BF5B3D;
}

/* all prompts */
div.nbinput.container div.prompt,
div.nboutput.container div.prompt {
    width: 4.5ex;
    padding-top: 5px;
    position: relative;
    user-select: none;
}

div.nbinput.container div.prompt > div,
div.nboutput.container div.prompt > div {
    position: absolute;
    right: 0;
    margin-right: 0.3ex;
}

@media (max-width: 540px) {
    div.nbinput.container div.prompt,
    div.nboutput.container div.prompt {
        width: unset;
        text-align: left;
        padding: 0.4em;
    }
    div.nboutput.container div.prompt.empty {
        padding: 0;
    }

    div.nbinput.container div.prompt > div,
    div.nboutput.container div.prompt > div {
        position: unset;
    }
}

/* disable scrollbars on prompts */
div.nbinput.container div.prompt pre,
div.nboutput.container div.prompt pre {
    overflow: hidden;
}

/* input/output area */
div.nbinput.container div.input_area,
div.nboutput.container div.output_area {
    -webkit-flex: 1;
    flex: 1;
    overflow: auto;
}
@media (max-width: 540px) {
    div.nbinput.container div.input_area,
    div.nboutput.container div.output_area {
        width: 100%;
    }
}

/* input area */
div.nbinput.container div.input_area {
    border: 1px solid #e0e0e0;
    border-radius: 2px;
    /*background: #f5f5f5;*/
}

/* override MathJax center alignment in output cells */
div.nboutput.container div[class*=MathJax] {
    text-align: left !important;
}

/* override sphinx.ext.imgmath center alignment in output cells */
div.nboutput.container div.math p {
    text-align: left;
}

/* standard error */
div.nboutput.container div.output_area.stderr {
    background: #fdd;
}

/* ANSI colors */
.ansi-black-fg { color: #3E424D; }
.ansi-black-bg { background-color: #3E424D; }
.ansi-black-intense-fg { color: #282C36; }
.ansi-black-intense-bg { background-color: #282C36; }
.ansi-red-fg { color: #E75C58; }
.ansi-red-bg { background-color: #E75C58; }
.ansi-red-intense-fg { color: #B22B31; }
.ansi-red-intense-bg { background-color: #B22B31; }
.ansi-green-fg { color: #00A250; }
.ansi-green-bg { background-color: #00A250; }
.ansi-green-intense-fg { color: #007427; }
.ansi-green-intense-bg { background-color: #007427; }
.ansi-yellow-fg { color: #DDB62B; }
.ansi-yellow-bg { background-color: #DDB62B; }
.ansi-yellow-intense-fg { color: #B27D12; }
.ansi-yellow-intense-bg { background-color: #B27D12; }
.ansi-blue-fg { color: #208FFB; }
.ansi-blue-bg { background-color: #208FFB; }
.ansi-blue-intense-fg { color: #0065CA; }
.ansi-blue-intense-bg { background-color: #0065CA; }
.ansi-magenta-fg { color: #D160C4; }
.ansi-magenta-bg { background-color: #D160C4; }
.ansi-magenta-intense-fg { color: #A03196; }
.ansi-magenta-intense-bg { background-color: #A03196; }
.ansi-cyan-fg { color: #60C6C8; }
.ansi-cyan-bg { background-color: #60C6C8; }
.ansi-cyan-intense-fg { color: #258F8F; }
.ansi-cyan-intense-bg { background-color: #258F8F; }
.ansi-white-fg { color: #C5C1B4; }
.ansi-white-bg { background-color: #C5C1B4; }
.ansi-white-intense-fg { color: #A1A6B2; }
.ansi-white-intense-bg { background-color: #A1A6B2; }

.ansi-default-inverse-fg { color: #FFFFFF; }
.ansi-default-inverse-bg { background-color: #000000; }

.ansi-bold { font-weight: bold; }
.ansi-underline { text-decoration: underline; }


div.nbinput.container div.input_area div[class*=highlight] > pre,
div.nboutput.container div.output_area div[class*=highlight] > pre,
div.nboutput.container div.output_area div[class*=highlight].math,
div.nboutput.container div.output_area.rendered_html,
div.nboutput.container div.output_area > div.output_javascript,
div.nboutput.container div.output_area:not(.rendered_html) > img{
    padding: 5px;
    margin: 0;
}

/* fix copybtn overflow problem in chromium (needed for 'sphinx_copybutton') */
div.nbinput.container div.input_area > div[class^='highlight'],
div.nboutput.container div.output_area > div[class^='highlight']{
    overflow-y: hidden;
}

/* hide copybtn icon on prompts (needed for 'sphinx_copybutton') */
.prompt .copybtn {
    display: none;
}

/* Some additional styling taken form the Jupyter notebook CSS */
.jp-RenderedHTMLCommon table,
div.rendered_html table {
  border: none;
  border-collapse: collapse;
  border-spacing: 0;
  color: black;
  font-size: 12px;
  table-layout: fixed;
}
.jp-RenderedHTMLCommon thead,
div.rendered_html thead {
  border-bottom: 1px solid black;
  vertical-align: bottom;
}
.jp-RenderedHTMLCommon tr,
.jp-RenderedHTMLCommon th,
.jp-RenderedHTMLCommon td,
div.rendered_html tr,
div.rendered_html th,
div.rendered_html td {
  text-align: right;
  vertical-align: middle;
  padding: 0.5em 0.5em;
  line-height: normal;
  white-space: normal;
  max-width: none;
  border: none;
}
.jp-RenderedHTMLCommon th,
div.rendered_html th {
  font-weight: bold;
}
.jp-RenderedHTMLCommon tbody tr:nth-child(odd),
div.rendered_html tbody tr:nth-child(odd) {
  background: #f5f5f5;
}
.jp-RenderedHTMLCommon tbody tr:hover,
div.rendered_html tbody tr:hover {
  background: rgba(66, 165, 245, 0.2);
}
</style>
<p><a class="reference external" href="https://colab.research.google.com/github/sagar87/scPCA/blob/main/docs/notebooks/kang.ipynb"><img alt="Open In Colab" src="https://colab.research.google.com/assets/colab-badge.svg" /></a></p>
<div class="admonition note"><p class="admonition-title"><p>Note</p>
</p><p><p>In order to run this notebook in Google Colab run the following cells. For best performance make sure that you run the notebook on a GPU instance, i.e. choose from the menu bar Runtime &gt; Change Runtime &gt; Hardware accelerator &gt; GPU.</p>
</p></div><div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[3]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span># Install scPCA + dependencies
!pip install --quiet scpca scikit-misc
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[2]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span># Download the Kang et. al. dataset
!wget https://www.huber.embl.de/users/harald/scpca/angelidis.h5ad /content/angelidis.h5ad
</pre></div>
</div>
</div>
<section id="Analysing-the-brain-cells-in-light-stimulated-mice">
<h1>Analysing the brain cells in light stimulated mice<a class="headerlink" href="#Analysing-the-brain-cells-in-light-stimulated-mice" title="Permalink to this heading">#</a></h1>
<p>In this notebook, we analyse the dataset from Angelidis et. al. in which the authors investigated lung cells of aging mice.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[54]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>import scpca as scp
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import scanpy as sc
import seaborn as sns
</pre></div>
</div>
</div>
<section id="Data-preprocessing">
<h2>Data preprocessing<a class="headerlink" href="#Data-preprocessing" title="Permalink to this heading">#</a></h2>
<p>We begin by importing the dataset with <code class="docutils literal notranslate"><span class="pre">scanpy</span></code>.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[57]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>data_path = &#39;/content/hrvatin.h5ad&#39;
adata = sc.read_h5ad(data_path)
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[58]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>adata
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[58]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
AnnData object with n_obs × n_vars = 14745 × 20908
    obs: &#39;barcode&#39;, &#39;mouse_id&#39;, &#39;age&#39;, &#39;cell_type&#39;, &#39;cell_id&#39;
    uns: &#39;X_name&#39;, &#39;celltype_column&#39;, &#39;condition_column&#39;, &#39;main_assay&#39;
</pre></div></div>
</div>
<p>Examining the obs field of the <code class="docutils literal notranslate"><span class="pre">adata</span></code> object reveals that the dataset includes two conditions (<code class="docutils literal notranslate"><span class="pre">age</span></code> column), corresponding to mice aged 3 and 24 months.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[4]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>adata.obs
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[4]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>barcode</th>
      <th>mouse_id</th>
      <th>age</th>
      <th>cell_type</th>
      <th>cell_id</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>muc3838:GCACTTTAGAAT</th>
      <td>GCACTTTAGAAT</td>
      <td>muc3838</td>
      <td>24m</td>
      <td>Ciliated_cells</td>
      <td>muc3838:GCACTTTAGAAT</td>
    </tr>
    <tr>
      <th>muc3838:TCCTGCTCCCTT</th>
      <td>TCCTGCTCCCTT</td>
      <td>muc3838</td>
      <td>24m</td>
      <td>Ciliated_cells</td>
      <td>muc3838:TCCTGCTCCCTT</td>
    </tr>
    <tr>
      <th>muc3838:TCCTGCTCCCTG</th>
      <td>TCCTGCTCCCTG</td>
      <td>muc3838</td>
      <td>24m</td>
      <td>Ciliated_cells</td>
      <td>muc3838:TCCTGCTCCCTG</td>
    </tr>
    <tr>
      <th>muc3838:TGCGAGTCTTCT</th>
      <td>TGCGAGTCTTCT</td>
      <td>muc3838</td>
      <td>24m</td>
      <td>Ciliated_cells</td>
      <td>muc3838:TGCGAGTCTTCT</td>
    </tr>
    <tr>
      <th>muc3838:TCCTGCTCCCTC</th>
      <td>TCCTGCTCCCTC</td>
      <td>muc3838</td>
      <td>24m</td>
      <td>Ciliated_cells</td>
      <td>muc3838:TCCTGCTCCCTC</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>muc4657:ATCAAGACAGTG</th>
      <td>ATCAAGACAGTG</td>
      <td>muc4657</td>
      <td>3m</td>
      <td>Type_2_pneumocytes</td>
      <td>muc4657:ATCAAGACAGTG</td>
    </tr>
    <tr>
      <th>muc4657:GACTGCGCATGG</th>
      <td>GACTGCGCATGG</td>
      <td>muc4657</td>
      <td>3m</td>
      <td>Ciliated_cells</td>
      <td>muc4657:GACTGCGCATGG</td>
    </tr>
    <tr>
      <th>muc4657:ATGACCGAATGT</th>
      <td>ATGACCGAATGT</td>
      <td>muc4657</td>
      <td>3m</td>
      <td>Type_2_pneumocytes</td>
      <td>muc4657:ATGACCGAATGT</td>
    </tr>
    <tr>
      <th>muc4657:GTGTTTGGACCG</th>
      <td>GTGTTTGGACCG</td>
      <td>muc4657</td>
      <td>3m</td>
      <td>Cd4+_T_cells</td>
      <td>muc4657:GTGTTTGGACCG</td>
    </tr>
    <tr>
      <th>muc4657:TGGCCGTGTAGC</th>
      <td>TGGCCGTGTAGC</td>
      <td>muc4657</td>
      <td>3m</td>
      <td>Goblet_cells</td>
      <td>muc4657:TGGCCGTGTAGC</td>
    </tr>
  </tbody>
</table>
<p>14745 rows × 5 columns</p>
</div></div>
</div>
<p>Before proceeding with the selection of highly variable genes, we filter out low-quality cells and red blood cells, as the latter are not representative of lung cell types.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[5]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>adata = adata[~adata.obs[&#39;cell_type&#39;].isin([&#39;low_quality_cells&#39;, &#39;red_blood_cells&#39;])]
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[6]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>adata.layers[&quot;counts&quot;] = adata.X.copy()
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
/tmp/ipykernel_657/1517723426.py:1: ImplicitModificationWarning: Setting element `.layers[&#39;counts&#39;]` of view, initializing view as actual.
  adata.layers[&#34;counts&#34;] = adata.X.copy()
</pre></div></div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[7]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>sc.pp.highly_variable_genes(
    adata,
    n_top_genes=2000,
    flavor=&quot;seurat_v3&quot;,
    subset=True,
    layer=&quot;counts&quot;
)
</pre></div>
</div>
</div>
</section>
<section id="Factor-analysis">
<h2>Factor analysis<a class="headerlink" href="#Factor-analysis" title="Permalink to this heading">#</a></h2>
<p>Next, we model the data using scPCA. One strategy is to treat the 3-month condition as the reference and capture gene expression shifts in old mice as deviations from this baseline. This can be specified using the design formula <code class="docutils literal notranslate"><span class="pre">~</span> <span class="pre">age</span></code>, similar to how one would define a model in a linear regression framework.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[8]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>m0 = scp.scPCA(
    adata,
    num_factors=10,
    layers_key=&#39;counts&#39;,
    loadings_formula=&#39;age&#39;,
    intercept_formula=&#39;1&#39;,
    seed=8999074
)
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[9]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>m0.fit()
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
  0%|          | 0/5000 [00:00&lt;?, ?it/s]/g/huber/users/voehring/conda/scpca_dev/lib/python3.12/site-packages/scpca/models/scpca.py:68: UserWarning: The use of `x.T` on tensors of dimension other than 2 to reverse their shape is deprecated and it will throw an error in a future release. Consider `x.mT` to transpose batches of matrices or `x.permute(*torch.arange(x.ndim - 1, -1, -1))` to reverse the dimensions of a tensor. (Triggered internally at /pytorch/aten/src/ATen/native/TensorShape.cpp:4413.)
  α_rna = deterministic(&#34;α_rna&#34;, (1 / α_rna_inv).T)
Epoch: 4990 | lr: 1.00E-02 | ELBO: 2913242 | Δ_10: 43149.50 | Best: 2838556: 100%|██████████| 5000/5000 [02:12&lt;00:00, 37.65it/s]
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[10]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>m0.fit(lr=0.001, num_epochs=10000)
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
Epoch: 14990 | lr: 1.00E-03 | ELBO: 2876858 | Δ_10: 37865.50 | Best: 2809824: 100%|██████████| 10000/10000 [04:08&lt;00:00, 40.29it/s]
</pre></div></div>
</div>
<p>Let’s store the result from the factorisation in the <code class="docutils literal notranslate"><span class="pre">adata</span></code> object.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[11]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>m0.mean_to_anndata(model_key=&#39;m0&#39;, num_samples=10, num_split=512)
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
Predicting z for obs 13824-14065.: 100%|██████████| 28/28 [00:02&lt;00:00, 13.78it/s]
</pre></div></div>
</div>
<p>If we now take a close look at the adata object we will find that the factors and loadings were added to the <code class="docutils literal notranslate"><span class="pre">obsm</span></code> and <code class="docutils literal notranslate"><span class="pre">varm</span></code> fields respectively (<code class="docutils literal notranslate"><span class="pre">X_m0</span></code> and <code class="docutils literal notranslate"><span class="pre">W_m0</span></code>). Meta data related to the model field are stored under <code class="docutils literal notranslate"><span class="pre">adata.uns[{model_key}]</span></code>.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[12]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>adata
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[12]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
AnnData object with n_obs × n_vars = 14066 × 2000
    obs: &#39;barcode&#39;, &#39;mouse_id&#39;, &#39;age&#39;, &#39;cell_type&#39;, &#39;cell_id&#39;
    var: &#39;highly_variable&#39;, &#39;highly_variable_rank&#39;, &#39;means&#39;, &#39;variances&#39;, &#39;variances_norm&#39;
    uns: &#39;X_name&#39;, &#39;celltype_column&#39;, &#39;condition_column&#39;, &#39;main_assay&#39;, &#39;hvg&#39;, &#39;m0&#39;
    obsm: &#39;X_m0&#39;
    varm: &#39;W_m0&#39;
    layers: &#39;counts&#39;
</pre></div></div>
</div>
<p>Let’s check our factor representation by getting a visual overview using UMAP. We use the UMAP helper function of scpca to accomplish that.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[13]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>scp.tl.umap(adata, &#39;X_m0&#39;)
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[14]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>sc.pl.embedding(adata, basis=&#39;X_m0_umap&#39;, color=[&#39;cell_type&#39;, &#39;age&#39;], ncols=1)
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_angelidis_26_0.png" src="../_images/notebooks_angelidis_26_0.png" />
</div>
</div>
<p>We now try to assess the main factors of the decomposition.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[15]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>scp.pl.factor_embedding(adata, model_key=&#39;m0&#39;, factor=[0, 1, 2, 3], basis=&#39;X_m0_umap&#39;, ncols=2)
plt.tight_layout()
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_angelidis_28_0.png" src="../_images/notebooks_angelidis_28_0.png" />
</div>
</div>
<section id="Factor-0">
<h3>Factor 0<a class="headerlink" href="#Factor-0" title="Permalink to this heading">#</a></h3>
<p>We begin by examining factor 0 more closely. This factor displays notably negative weights in myeloid cells, including alveolar macrophages and monocytes, and positive weights in club and goblet cells.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[26]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>scp.pl.factor_strip(adata, model_key=&#39;m0&#39;, factor=[0], cluster_key=&#39;cell_type&#39;, highlight=[&#39;Alveolar_macrophage&#39;, &#39;Club_cells&#39;], s=2)
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[26]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
&lt;Axes: title={&#39;center&#39;: &#39;Factor 0&#39;}, xlabel=&#39;cluster&#39;, ylabel=&#39;Factor weight&#39;&gt;
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_angelidis_30_1.png" src="../_images/notebooks_angelidis_30_1.png" />
</div>
</div>
<p>Next we take a close look at the <em>negative</em> loading weights of factor 0 to learn more about the expression patterns in the myeloid cells of this dataset. We can extract those from the <code class="docutils literal notranslate"><span class="pre">adata</span></code> object using the <code class="docutils literal notranslate"><span class="pre">state_loadings</span></code> function. Since the factor weights are negative for the myeloid cells, we use the argument <code class="docutils literal notranslate"><span class="pre">lowest=20</span></code> and <code class="docutils literal notranslate"><span class="pre">highest=0</span></code> to extract the genes with the largest negative loading weights.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[34]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>top_weights_fac0 = scp.tl.state_loadings(adata, &#39;m0&#39;, &#39;Intercept&#39;, 0, highest=0, lowest=20)
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[35]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>top_weights_fac0
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[35]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>gene</th>
      <th>magnitude</th>
      <th>weight</th>
      <th>type</th>
      <th>state</th>
      <th>factor</th>
      <th>index</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>Abcg1</td>
      <td>8.844831</td>
      <td>-8.844831</td>
      <td>lowest</td>
      <td>0</td>
      <td>0</td>
      <td>73</td>
    </tr>
    <tr>
      <th>1</th>
      <td>Bcl2a1a</td>
      <td>8.854088</td>
      <td>-8.854088</td>
      <td>lowest</td>
      <td>0</td>
      <td>0</td>
      <td>162</td>
    </tr>
    <tr>
      <th>2</th>
      <td>Mpeg1</td>
      <td>9.212790</td>
      <td>-9.212790</td>
      <td>lowest</td>
      <td>0</td>
      <td>0</td>
      <td>1053</td>
    </tr>
    <tr>
      <th>3</th>
      <td>Marco</td>
      <td>9.213015</td>
      <td>-9.213015</td>
      <td>lowest</td>
      <td>0</td>
      <td>0</td>
      <td>1020</td>
    </tr>
    <tr>
      <th>4</th>
      <td>1100001G20Rik</td>
      <td>9.315811</td>
      <td>-9.315811</td>
      <td>lowest</td>
      <td>0</td>
      <td>0</td>
      <td>0</td>
    </tr>
    <tr>
      <th>5</th>
      <td>Atp6v0d2</td>
      <td>9.373060</td>
      <td>-9.373060</td>
      <td>lowest</td>
      <td>0</td>
      <td>0</td>
      <td>144</td>
    </tr>
    <tr>
      <th>6</th>
      <td>Ltc4s</td>
      <td>9.418015</td>
      <td>-9.418015</td>
      <td>lowest</td>
      <td>0</td>
      <td>0</td>
      <td>994</td>
    </tr>
    <tr>
      <th>7</th>
      <td>Fabp1</td>
      <td>9.497688</td>
      <td>-9.497688</td>
      <td>lowest</td>
      <td>0</td>
      <td>0</td>
      <td>542</td>
    </tr>
    <tr>
      <th>8</th>
      <td>Clec4n</td>
      <td>9.505602</td>
      <td>-9.505602</td>
      <td>lowest</td>
      <td>0</td>
      <td>0</td>
      <td>368</td>
    </tr>
    <tr>
      <th>9</th>
      <td>Mrc1</td>
      <td>9.583654</td>
      <td>-9.583654</td>
      <td>lowest</td>
      <td>0</td>
      <td>0</td>
      <td>1055</td>
    </tr>
    <tr>
      <th>10</th>
      <td>Ear1</td>
      <td>9.637006</td>
      <td>-9.637006</td>
      <td>lowest</td>
      <td>0</td>
      <td>0</td>
      <td>496</td>
    </tr>
    <tr>
      <th>11</th>
      <td>Bcl2a1d</td>
      <td>9.889424</td>
      <td>-9.889424</td>
      <td>lowest</td>
      <td>0</td>
      <td>0</td>
      <td>164</td>
    </tr>
    <tr>
      <th>12</th>
      <td>Clec4a3</td>
      <td>10.041142</td>
      <td>-10.041142</td>
      <td>lowest</td>
      <td>0</td>
      <td>0</td>
      <td>363</td>
    </tr>
    <tr>
      <th>13</th>
      <td>Fcer1g</td>
      <td>10.077384</td>
      <td>-10.077384</td>
      <td>lowest</td>
      <td>0</td>
      <td>0</td>
      <td>573</td>
    </tr>
    <tr>
      <th>14</th>
      <td>Tyrobp</td>
      <td>10.141203</td>
      <td>-10.141203</td>
      <td>lowest</td>
      <td>0</td>
      <td>0</td>
      <td>1513</td>
    </tr>
    <tr>
      <th>15</th>
      <td>Cybb</td>
      <td>10.273161</td>
      <td>-10.273161</td>
      <td>lowest</td>
      <td>0</td>
      <td>0</td>
      <td>435</td>
    </tr>
    <tr>
      <th>16</th>
      <td>Ccl6</td>
      <td>10.514906</td>
      <td>-10.514906</td>
      <td>lowest</td>
      <td>0</td>
      <td>0</td>
      <td>252</td>
    </tr>
    <tr>
      <th>17</th>
      <td>Chi3l3</td>
      <td>10.604345</td>
      <td>-10.604345</td>
      <td>lowest</td>
      <td>0</td>
      <td>0</td>
      <td>340</td>
    </tr>
    <tr>
      <th>18</th>
      <td>Ear2</td>
      <td>10.699730</td>
      <td>-10.699730</td>
      <td>lowest</td>
      <td>0</td>
      <td>0</td>
      <td>498</td>
    </tr>
    <tr>
      <th>19</th>
      <td>Ctss</td>
      <td>10.881914</td>
      <td>-10.881914</td>
      <td>lowest</td>
      <td>0</td>
      <td>0</td>
      <td>420</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
<p>As expected the genes are largely associated with myeloid cells, particularly macrophages, and more specifically, alveolar macrophages or monocyte-derived macrophages:</p>
<ul class="simple">
<li><p>Strong indicators of macrophages / myeloid lineage: Mpeg1, Clec4a3, Clec4n, Fcer1g, Tyrobp, Cybb, Ctss – general myeloid/macrophage markers.</p></li>
<li><p>Marco, Mrc1 (CD206) – alveolar macrophage markers; Marco is specific to tissue-resident macrophages like alveolar macrophages.</p></li>
<li><p>Chi3l3 (Ym1), Ear1, Ear2, Ccl6 – often upregulated in alternatively activated (M2-like) macrophages in mice.</p></li>
<li><p>Bcl2a1a, Bcl2a1d – anti-apoptotic genes, enriched in immune cells including myeloid cells.</p></li>
<li><p>Atp6v0d2 – involved in osteoclast/macrophage function.</p></li>
<li><p>Ltc4s – leukotriene pathway, expressed in myeloid-derived cells.</p></li>
<li><p>Abcg1 – involved in lipid homeostasis, enriched in alveolar macrophages.</p></li>
<li><p>Tyrobp – adapter protein for many immune receptors, common in microglia/macrophages.</p></li>
</ul>
<p>We next ask the question how the factor 0 changed across conditions. To this end we can compute the loading differences between factor 0 loading weigths between the 24m and 3m condition.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[41]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>_ = scp.pl.loadings_state(adata, &#39;m0&#39;, [&#39;Intercept&#39;, &#39;age[T.24m]&#39;], 0, sign=-1, highest=10, lowest=10, width=4, height=3.5)
plt.tight_layout()
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_angelidis_36_0.png" src="../_images/notebooks_angelidis_36_0.png" />
</div>
</div>
<p>Interestingly we can see here see the downregulation of Ltc4s, Ear1, Ear2 – These are typically expressed in tissue-resident lung macrophages. Their downregulation suggests a shift in macrophage state (e.g., toward a more pro-inflammatory phenotype or loss of identity).</p>
<p>The set of upregulated genes appears to reflect a shift in macrophage state—potentially toward a more inflammatory, monocyte-derived, or activated phenotype, possibly related to aging or stress in the lung environment. This is especially supported by the upregulation of:</p>
<ul class="simple">
<li><p>Nr4a1: Transcription factor critical for the development of non-classical monocytes; also involved in anti-inflammatory responses and macrophage self-renewal. Upregulation may indicate a monocyte-to-macrophage transition or stress adaptation.</p></li>
<li><p>Apoe, Apoc2: Apolipoproteins associated with lipid metabolism and foam-cell–like macrophage states; commonly upregulated in aging, inflammation, or lipid-rich environments (e.g., alveolar space). These are known markers of activated macrophages in chronic inflammation and neurodegeneration, too.</p></li>
</ul>
</section>
<section id="Factor-1">
<h3>Factor 1<a class="headerlink" href="#Factor-1" title="Permalink to this heading">#</a></h3>
<p>Next we look at factor 1. Factor 1 exhibits high factor values in Ciliated Cells. Let’s also have a look factor 1s corresponding loading weights.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[31]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>top_weights_fac1 = scp.tl.state_loadings(adata, &#39;m0&#39;, &#39;Intercept&#39;, 1, highest=20)
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[33]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>top_weights_fac1.head()
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[33]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>gene</th>
      <th>magnitude</th>
      <th>weight</th>
      <th>type</th>
      <th>state</th>
      <th>factor</th>
      <th>index</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>Fam183b</td>
      <td>8.030411</td>
      <td>8.030411</td>
      <td>highest</td>
      <td>0</td>
      <td>1</td>
      <td>554</td>
    </tr>
    <tr>
      <th>1</th>
      <td>1700016K19Rik</td>
      <td>7.806339</td>
      <td>7.806339</td>
      <td>highest</td>
      <td>0</td>
      <td>1</td>
      <td>15</td>
    </tr>
    <tr>
      <th>2</th>
      <td>Mlf1</td>
      <td>7.700116</td>
      <td>7.700116</td>
      <td>highest</td>
      <td>0</td>
      <td>1</td>
      <td>1041</td>
    </tr>
    <tr>
      <th>3</th>
      <td>Tmem212</td>
      <td>7.574149</td>
      <td>7.574149</td>
      <td>highest</td>
      <td>0</td>
      <td>1</td>
      <td>1443</td>
    </tr>
    <tr>
      <th>4</th>
      <td>1700024G13Rik</td>
      <td>7.569454</td>
      <td>7.569454</td>
      <td>highest</td>
      <td>0</td>
      <td>1</td>
      <td>17</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[35]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>ax = sns.barplot(data=top_weights_fac1, x=&#39;gene&#39;, y=&#39;weight&#39;)
ax.tick_params(axis=&#39;x&#39;, rotation=90)
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_angelidis_41_0.png" src="../_images/notebooks_angelidis_41_0.png" />
</div>
</div>
<p>These genes—like <em>CCDC39</em>, <em>Drc1</em>, <em>Dynlrb2</em>, plus several coiled‑coil domain proteins (<em>CCDC113</em>, <em>CCDC153</em>, etc.)—are classic markers of motile cilia structures: they play key roles in axonemal assembly, dynein arm formation, and ciliary motility. For example, <em>CCDC39</em> is essential for the dynein regulatory complex and inner dynein arms, and <em>DRC1</em> also has a similar role in ciliary beat regulation.</p>
<p>We next ask the question how the factor 1 changed across conditions. To this end we can compute the loading differences between factor 1 loading weigths from the 24m and 3m condition.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[44]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>_ = scp.pl.loadings_state(adata, &#39;m0&#39;, [&#39;Intercept&#39;, &#39;age[T.24m]&#39;], 1, highest=10, width=4, height=3.5)
plt.tight_layout()
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_angelidis_44_0.png" src="../_images/notebooks_angelidis_44_0.png" />
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[50]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>top_loading_diff_1 = scp.tl.state_diff(adata, &#39;m0&#39;,  [&#39;Intercept&#39;, &#39;age[T.24m]&#39;], 1, highest=20)
</pre></div>
</div>
</div>
<p>These genes, when considered collectively, point to a signature of metabolically active, detoxifying epithelial cells—possibly from liver or lung tissue. They combine:</p>
<ul class="simple">
<li><p>Detox enzymes and transporters (P450s, Sult1d1, Slc16a11)</p></li>
<li><p>Epithelial structural/barrier proteins (Krt8, Cldn3, WFDC2)</p></li>
<li><p>Stress response and regulatory factors (Nupr1, Tiam1, Wif1, Dkkl1)</p></li>
</ul>
<p>Together these results indicate that scPCAs inferred loading weight differences indicate that ciliated cells shift toward a more metabollically active / stress respond state in old mice.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>
</pre></div>
</div>
</div>
</section>
</section>
</section>


                </article>
              

              
              
              
              
                <footer class="prev-next-footer">
                  <!-- Previous / next buttons -->
<div class="prev-next-area">
    <a class="left-prev"
       href="pancreas.html"
       title="previous page">
      <i class="fa-solid fa-angle-left"></i>
      <div class="prev-next-info">
        <p class="prev-next-subtitle">previous</p>
        <p class="prev-next-title">Integrating the human pancreas dataset</p>
      </div>
    </a>
    <a class="right-next"
       href="hrvatin.html"
       title="next page">
      <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">Analysing the lung cell populations in aging mice</p>
      </div>
      <i class="fa-solid fa-angle-right"></i>
    </a>
</div>
                </footer>
              
            </div>
            
            
              
                <div class="bd-sidebar-secondary bd-toc"><div class="sidebar-secondary-items sidebar-secondary__inner">

  <div class="sidebar-secondary-item">
  <div class="page-toc tocsection onthispage">
    <i class="fa-solid fa-list"></i> Contents
  </div>
  <nav class="bd-toc-nav page-toc">
    <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#Data-preprocessing">Data preprocessing</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#Factor-analysis">Factor analysis</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#Factor-0">Factor 0</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#Factor-1">Factor 1</a></li>
</ul>
</li>
</ul>
  </nav></div>

</div></div>
              
            
          </div>
          <footer class="bd-footer-content">
            
<div class="bd-footer-content__inner container">
  
  <div class="footer-item">
    
<p class="component-author">
By Harald Vohringer
</p>

  </div>
  
  <div class="footer-item">
    
  <p class="copyright">
    
      © Copyright 2023, Harald Vohringer.
      <br/>
    
  </p>

  </div>
  
  <div class="footer-item">
    
  </div>
  
  <div class="footer-item">
    
  </div>
  
</div>
          </footer>
        

      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="../_static/scripts/bootstrap.js?digest=927b94d3fcb96560df09"></script>
<script src="../_static/scripts/pydata-sphinx-theme.js?digest=927b94d3fcb96560df09"></script>

  <footer class="bd-footer">
  </footer>
  </body>
</html>