

<!DOCTYPE html>


<html lang="en" >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

    <title>Analysing the lung cell populations in aging mice &#8212; scPCA 0.3.3 documentation</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "light";
  </script>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="../_static/styles/theme.css?digest=927b94d3fcb96560df09" rel="stylesheet" />
<link href="../_static/styles/bootstrap.css?digest=927b94d3fcb96560df09" rel="stylesheet" />
<link href="../_static/styles/pydata-sphinx-theme.css?digest=927b94d3fcb96560df09" rel="stylesheet" />

  
  <link href="../_static/vendor/fontawesome/6.1.2/css/all.min.css?digest=927b94d3fcb96560df09" rel="stylesheet" />
  <link rel="preload" as="font" type="font/woff2" crossorigin href="../_static/vendor/fontawesome/6.1.2/webfonts/fa-solid-900.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../_static/vendor/fontawesome/6.1.2/webfonts/fa-brands-400.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../_static/vendor/fontawesome/6.1.2/webfonts/fa-regular-400.woff2" />

    <link rel="stylesheet" type="text/css" href="../_static/pygments.css" />
    <link rel="stylesheet" href="../_static/styles/sphinx-book-theme.css?digest=14f4ca6b54d191a8c7657f6c759bf11a5fb86285" type="text/css" />
  
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="../_static/scripts/bootstrap.js?digest=927b94d3fcb96560df09" />
<link rel="preload" as="script" href="../_static/scripts/pydata-sphinx-theme.js?digest=927b94d3fcb96560df09" />
  <script src="../_static/vendor/fontawesome/6.1.2/js/all.min.js?digest=927b94d3fcb96560df09"></script>

    <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js"></script>
    <script src="../_static/jquery.js"></script>
    <script src="../_static/underscore.js"></script>
    <script src="../_static/_sphinx_javascript_frameworks_compat.js"></script>
    <script src="../_static/doctools.js"></script>
    <script src="../_static/sphinx_highlight.js"></script>
    <script src="../_static/scripts/sphinx-book-theme.js?digest=5a5c038af52cf7bc1a1ec88eea08e6366ee68824"></script>
    <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script>window.MathJax = {"tex": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true}, "options": {"ignoreHtmlClass": "tex2jax_ignore|mathjax_ignore|document", "processHtmlClass": "tex2jax_process|mathjax_process|math|output_area"}}</script>
    <script defer="defer" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    <script>DOCUMENTATION_OPTIONS.pagename = 'notebooks/hrvatin';</script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="prev" title="Analysing the brain cells in light stimulated mice" href="angelidis.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <a class="skip-link" href="#main-content">Skip to main content</a>
  
  <div id="pst-scroll-pixel-helper"></div>

  
  <button type="button" class="btn rounded-pill" id="pst-back-to-top">
    <i class="fa-solid fa-arrow-up"></i>
    Back to top
  </button>

  
  <input type="checkbox"
          class="sidebar-toggle"
          name="__primary"
          id="__primary"/>
  <label class="overlay overlay-primary" for="__primary"></label>
  
  <input type="checkbox"
          class="sidebar-toggle"
          name="__secondary"
          id="__secondary"/>
  <label class="overlay overlay-secondary" for="__secondary"></label>
  
  <div class="search-button__wrapper">
    <div class="search-button__overlay"></div>
    <div class="search-button__search-container">
<form class="bd-search d-flex align-items-center"
      action="../search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         id="search-input"
         placeholder="Search..."
         aria-label="Search..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form></div>
  </div>
  
    <nav class="bd-header navbar navbar-expand-lg bd-navbar">
    </nav>
  
  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      <div class="bd-sidebar-primary bd-sidebar">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
    
    
  </div>
  
    <div class="sidebar-primary-items__start sidebar-primary__section">
        <div class="sidebar-primary-item">
  

<a class="navbar-brand logo" href="../index.html">
  
  
  
  
  
    <p class="title logo__title">scPCA 0.3.3 documentation</p>
  
</a></div>
        <div class="sidebar-primary-item"><nav class="bd-links" id="bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item navbar-nav active">
        <p aria-level="2" class="caption" role="heading"><span class="caption-text">API</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../models.html">Models</a></li>
<li class="toctree-l1"><a class="reference internal" href="../tools.html">Tools</a></li>
<li class="toctree-l1"><a class="reference internal" href="../plots.html">Plotting</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Tutorials</span></p>
<ul class="current nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="kang.html">Analysing interferon stimulated PBMCs using scPCA</a></li>
<li class="toctree-l1"><a class="reference internal" href="pancreas.html">Integrating the human pancreas dataset</a></li>
<li class="toctree-l1"><a class="reference internal" href="angelidis.html">Analysing the brain cells in light stimulated mice</a></li>
<li class="toctree-l1 current active"><a class="current reference internal" href="#">Analysing the lung cell populations in aging mice</a></li>
</ul>

    </div>
</nav></div>
    </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
  </div>
  
  <div id="rtd-footer-container"></div>


      </div>
      
      <main id="main-content" class="bd-main">
        
        

<div class="sbt-scroll-pixel-helper"></div>

          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item"><label class="sidebar-toggle primary-toggle btn btn-sm" for="__primary" title="Toggle primary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
  <span class="fa-solid fa-bars"></span>
</label></div>
      
    </div>
  
  
    <div class="header-article-items__end">
      
        <div class="header-article-item">

<div class="article-header-buttons">





<div class="dropdown dropdown-download-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Download this page">
    <i class="fas fa-download"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="../_sources/notebooks/hrvatin.ipynb" target="_blank"
   class="btn btn-sm btn-download-source-button dropdown-item"
   title="Download source file"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file"></i>
  </span>
<span class="btn__text-container">.ipynb</span>
</a>
</li>
      
      
      
      
      <li>
<button onclick="window.print()"
  class="btn btn-sm btn-download-pdf-button dropdown-item"
  title="Print to PDF"
  data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file-pdf"></i>
  </span>
<span class="btn__text-container">.pdf</span>
</button>
</li>
      
  </ul>
</div>




<button onclick="toggleFullScreen()"
  class="btn btn-sm btn-fullscreen-button"
  title="Fullscreen mode"
  data-bs-placement="bottom" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>


<script>
document.write(`
  <button class="btn btn-sm navbar-btn theme-switch-button" title="light/dark" aria-label="light/dark" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="theme-switch nav-link" data-mode="light"><i class="fa-solid fa-sun fa-lg"></i></span>
    <span class="theme-switch nav-link" data-mode="dark"><i class="fa-solid fa-moon fa-lg"></i></span>
    <span class="theme-switch nav-link" data-mode="auto"><i class="fa-solid fa-circle-half-stroke fa-lg"></i></span>
  </button>
`);
</script>

<script>
document.write(`
  <button class="btn btn-sm navbar-btn search-button search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass fa-lg"></i>
  </button>
`);
</script>
<label class="sidebar-toggle secondary-toggle btn btn-sm" for="__secondary"title="Toggle secondary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="fa-solid fa-list"></span>
</label>
</div></div>
      
    </div>
  
</div>
</div>
              
              

<div id="jb-print-docs-body" class="onlyprint">
    <h1>Analysing the lung cell populations in aging mice</h1>
    <!-- Table of contents -->
    <div id="print-main-content">
        <div id="jb-print-toc">
            
            <div>
                <h2> Contents </h2>
            </div>
            <nav aria-label="Page">
                <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#Data-preprocessing">Data preprocessing</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#Factor-analysis">Factor analysis</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#Factor-0">Factor 0</a></li>
</ul>
</li>
</ul>
            </nav>
        </div>
    </div>
</div>

              
                
<div id="searchbox"></div>
                <article class="bd-article" role="main">
                  
  
<style>
/* CSS for nbsphinx extension */

/* remove conflicting styling from Sphinx themes */
div.nbinput.container div.prompt *,
div.nboutput.container div.prompt *,
div.nbinput.container div.input_area pre,
div.nboutput.container div.output_area pre,
div.nbinput.container div.input_area .highlight,
div.nboutput.container div.output_area .highlight {
    border: none;
    padding: 0;
    margin: 0;
    box-shadow: none;
}

div.nbinput.container > div[class*=highlight],
div.nboutput.container > div[class*=highlight] {
    margin: 0;
}

div.nbinput.container div.prompt *,
div.nboutput.container div.prompt * {
    background: none;
}

div.nboutput.container div.output_area .highlight,
div.nboutput.container div.output_area pre {
    background: unset;
}

div.nboutput.container div.output_area div.highlight {
    color: unset;  /* override Pygments text color */
}

/* avoid gaps between output lines */
div.nboutput.container div[class*=highlight] pre {
    line-height: normal;
}

/* input/output containers */
div.nbinput.container,
div.nboutput.container {
    display: -webkit-flex;
    display: flex;
    align-items: flex-start;
    margin: 0;
    width: 100%;
}
@media (max-width: 540px) {
    div.nbinput.container,
    div.nboutput.container {
        flex-direction: column;
    }
}

/* input container */
div.nbinput.container {
    padding-top: 5px;
}

/* last container */
div.nblast.container {
    padding-bottom: 5px;
}

/* input prompt */
div.nbinput.container div.prompt pre {
    color: #307FC1;
}

/* output prompt */
div.nboutput.container div.prompt pre {
    color: #BF5B3D;
}

/* all prompts */
div.nbinput.container div.prompt,
div.nboutput.container div.prompt {
    width: 4.5ex;
    padding-top: 5px;
    position: relative;
    user-select: none;
}

div.nbinput.container div.prompt > div,
div.nboutput.container div.prompt > div {
    position: absolute;
    right: 0;
    margin-right: 0.3ex;
}

@media (max-width: 540px) {
    div.nbinput.container div.prompt,
    div.nboutput.container div.prompt {
        width: unset;
        text-align: left;
        padding: 0.4em;
    }
    div.nboutput.container div.prompt.empty {
        padding: 0;
    }

    div.nbinput.container div.prompt > div,
    div.nboutput.container div.prompt > div {
        position: unset;
    }
}

/* disable scrollbars on prompts */
div.nbinput.container div.prompt pre,
div.nboutput.container div.prompt pre {
    overflow: hidden;
}

/* input/output area */
div.nbinput.container div.input_area,
div.nboutput.container div.output_area {
    -webkit-flex: 1;
    flex: 1;
    overflow: auto;
}
@media (max-width: 540px) {
    div.nbinput.container div.input_area,
    div.nboutput.container div.output_area {
        width: 100%;
    }
}

/* input area */
div.nbinput.container div.input_area {
    border: 1px solid #e0e0e0;
    border-radius: 2px;
    /*background: #f5f5f5;*/
}

/* override MathJax center alignment in output cells */
div.nboutput.container div[class*=MathJax] {
    text-align: left !important;
}

/* override sphinx.ext.imgmath center alignment in output cells */
div.nboutput.container div.math p {
    text-align: left;
}

/* standard error */
div.nboutput.container div.output_area.stderr {
    background: #fdd;
}

/* ANSI colors */
.ansi-black-fg { color: #3E424D; }
.ansi-black-bg { background-color: #3E424D; }
.ansi-black-intense-fg { color: #282C36; }
.ansi-black-intense-bg { background-color: #282C36; }
.ansi-red-fg { color: #E75C58; }
.ansi-red-bg { background-color: #E75C58; }
.ansi-red-intense-fg { color: #B22B31; }
.ansi-red-intense-bg { background-color: #B22B31; }
.ansi-green-fg { color: #00A250; }
.ansi-green-bg { background-color: #00A250; }
.ansi-green-intense-fg { color: #007427; }
.ansi-green-intense-bg { background-color: #007427; }
.ansi-yellow-fg { color: #DDB62B; }
.ansi-yellow-bg { background-color: #DDB62B; }
.ansi-yellow-intense-fg { color: #B27D12; }
.ansi-yellow-intense-bg { background-color: #B27D12; }
.ansi-blue-fg { color: #208FFB; }
.ansi-blue-bg { background-color: #208FFB; }
.ansi-blue-intense-fg { color: #0065CA; }
.ansi-blue-intense-bg { background-color: #0065CA; }
.ansi-magenta-fg { color: #D160C4; }
.ansi-magenta-bg { background-color: #D160C4; }
.ansi-magenta-intense-fg { color: #A03196; }
.ansi-magenta-intense-bg { background-color: #A03196; }
.ansi-cyan-fg { color: #60C6C8; }
.ansi-cyan-bg { background-color: #60C6C8; }
.ansi-cyan-intense-fg { color: #258F8F; }
.ansi-cyan-intense-bg { background-color: #258F8F; }
.ansi-white-fg { color: #C5C1B4; }
.ansi-white-bg { background-color: #C5C1B4; }
.ansi-white-intense-fg { color: #A1A6B2; }
.ansi-white-intense-bg { background-color: #A1A6B2; }

.ansi-default-inverse-fg { color: #FFFFFF; }
.ansi-default-inverse-bg { background-color: #000000; }

.ansi-bold { font-weight: bold; }
.ansi-underline { text-decoration: underline; }


div.nbinput.container div.input_area div[class*=highlight] > pre,
div.nboutput.container div.output_area div[class*=highlight] > pre,
div.nboutput.container div.output_area div[class*=highlight].math,
div.nboutput.container div.output_area.rendered_html,
div.nboutput.container div.output_area > div.output_javascript,
div.nboutput.container div.output_area:not(.rendered_html) > img{
    padding: 5px;
    margin: 0;
}

/* fix copybtn overflow problem in chromium (needed for 'sphinx_copybutton') */
div.nbinput.container div.input_area > div[class^='highlight'],
div.nboutput.container div.output_area > div[class^='highlight']{
    overflow-y: hidden;
}

/* hide copybtn icon on prompts (needed for 'sphinx_copybutton') */
.prompt .copybtn {
    display: none;
}

/* Some additional styling taken form the Jupyter notebook CSS */
.jp-RenderedHTMLCommon table,
div.rendered_html table {
  border: none;
  border-collapse: collapse;
  border-spacing: 0;
  color: black;
  font-size: 12px;
  table-layout: fixed;
}
.jp-RenderedHTMLCommon thead,
div.rendered_html thead {
  border-bottom: 1px solid black;
  vertical-align: bottom;
}
.jp-RenderedHTMLCommon tr,
.jp-RenderedHTMLCommon th,
.jp-RenderedHTMLCommon td,
div.rendered_html tr,
div.rendered_html th,
div.rendered_html td {
  text-align: right;
  vertical-align: middle;
  padding: 0.5em 0.5em;
  line-height: normal;
  white-space: normal;
  max-width: none;
  border: none;
}
.jp-RenderedHTMLCommon th,
div.rendered_html th {
  font-weight: bold;
}
.jp-RenderedHTMLCommon tbody tr:nth-child(odd),
div.rendered_html tbody tr:nth-child(odd) {
  background: #f5f5f5;
}
.jp-RenderedHTMLCommon tbody tr:hover,
div.rendered_html tbody tr:hover {
  background: rgba(66, 165, 245, 0.2);
}
</style>
<p><a class="reference external" href="https://colab.research.google.com/github/sagar87/scPCA/blob/main/docs/notebooks/kang.ipynb"><img alt="Open In Colab" src="https://colab.research.google.com/assets/colab-badge.svg" /></a></p>
<div class="admonition note"><p class="admonition-title"><p>Note</p>
</p><p><p>In order to run this notebook in Google Colab run the following cells. For best performance make sure that you run the notebook on a GPU instance, i.e. choose from the menu bar Runtime &gt; Change Runtime &gt; Hardware accelerator &gt; GPU.</p>
</p></div><div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[3]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span># Install scPCA + dependencies
!pip install --quiet scpca scikit-misc
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[2]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span># Download the Kang et. al. dataset
!wget https://www.huber.embl.de/users/harald/scpca/hrvatin.h5ad /content/hrvatin.h5ad
</pre></div>
</div>
</div>
<section id="Analysing-the-lung-cell-populations-in-aging-mice">
<h1>Analysing the lung cell populations in aging mice<a class="headerlink" href="#Analysing-the-lung-cell-populations-in-aging-mice" title="Permalink to this heading">#</a></h1>
<p>In this notebook, we analyse the dataset from Hrvatin et. al. in which the authors investigated light stimulated brain cells after various time points.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[1]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>import scpca as scp
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import scanpy as sc
import seaborn as sns
</pre></div>
</div>
</div>
<section id="Data-preprocessing">
<h2>Data preprocessing<a class="headerlink" href="#Data-preprocessing" title="Permalink to this heading">#</a></h2>
<p>We begin by importing the dataset with <code class="docutils literal notranslate"><span class="pre">scanpy</span></code>.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[2]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>data_path = &#39;/content/hrvatin.h5ad&#39;
adata = sc.read_h5ad(data_path)
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[3]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>adata
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[3]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
AnnData object with n_obs × n_vars = 65539 × 25187
    obs: &#39;Cell&#39;, &#39;stim&#39;, &#39;sample&#39;, &#39;maintype&#39;, &#39;celltype&#39;, &#39;subtype&#39;
    uns: &#39;X_name&#39;, &#39;celltype_column&#39;, &#39;condition_column&#39;, &#39;main_assay&#39;
</pre></div></div>
</div>
<p>Examining the obs field of the <code class="docutils literal notranslate"><span class="pre">adata</span></code> object reveals that the dataset includes three different time points (0h, 1h and 4h) in the <code class="docutils literal notranslate"><span class="pre">stim</span></code> column.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[4]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>adata.obs
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[4]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Cell</th>
      <th>stim</th>
      <th>sample</th>
      <th>maintype</th>
      <th>celltype</th>
      <th>subtype</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>x2_35_0_bc0013</th>
      <td>x2_35_0_bc0013</td>
      <td>0h</td>
      <td>B1_1_0h</td>
      <td>Excitatory</td>
      <td>ExcL4</td>
      <td>ExcL4_3</td>
    </tr>
    <tr>
      <th>x2_35_0_bc0014</th>
      <td>x2_35_0_bc0014</td>
      <td>0h</td>
      <td>B1_1_0h</td>
      <td>Excitatory</td>
      <td>ExcL5_3</td>
      <td>NA</td>
    </tr>
    <tr>
      <th>x2_35_0_bc0016</th>
      <td>x2_35_0_bc0016</td>
      <td>0h</td>
      <td>B1_1_0h</td>
      <td>Excitatory</td>
      <td>ExcL5_2</td>
      <td>NA</td>
    </tr>
    <tr>
      <th>x2_35_0_bc0017</th>
      <td>x2_35_0_bc0017</td>
      <td>0h</td>
      <td>B1_1_0h</td>
      <td>Excitatory</td>
      <td>ExcL6</td>
      <td>NA</td>
    </tr>
    <tr>
      <th>x2_35_0_bc0018</th>
      <td>x2_35_0_bc0018</td>
      <td>0h</td>
      <td>B1_1_0h</td>
      <td>Excitatory</td>
      <td>ExcL5_3</td>
      <td>NA</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>x2_98_4_2_2_bc2789</th>
      <td>x2_98_4_2_2_bc2789</td>
      <td>4h</td>
      <td>B7_23_4h_B</td>
      <td>NA</td>
      <td>NA</td>
      <td>NA</td>
    </tr>
    <tr>
      <th>x2_98_4_2_2_bc2805</th>
      <td>x2_98_4_2_2_bc2805</td>
      <td>4h</td>
      <td>B7_23_4h_B</td>
      <td>Microglia</td>
      <td>Micro_2</td>
      <td>NA</td>
    </tr>
    <tr>
      <th>x2_98_4_2_2_bc2823</th>
      <td>x2_98_4_2_2_bc2823</td>
      <td>4h</td>
      <td>B7_23_4h_B</td>
      <td>NA</td>
      <td>NA</td>
      <td>NA</td>
    </tr>
    <tr>
      <th>x2_98_4_2_2_bc2857</th>
      <td>x2_98_4_2_2_bc2857</td>
      <td>4h</td>
      <td>B7_23_4h_B</td>
      <td>NA</td>
      <td>NA</td>
      <td>NA</td>
    </tr>
    <tr>
      <th>x2_98_4_2_2_bc2997</th>
      <td>x2_98_4_2_2_bc2997</td>
      <td>4h</td>
      <td>B7_23_4h_B</td>
      <td>NA</td>
      <td>NA</td>
      <td>NA</td>
    </tr>
  </tbody>
</table>
<p>65539 rows × 6 columns</p>
</div></div>
</div>
<p>Before proceeding with the selection of highly variable genes, we filter out low-quality (“NA”) cells.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[5]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>adata = adata[~adata.obs[&#39;maintype&#39;].isin([&#39;NA&#39;])]
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[6]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>adata.layers[&quot;counts&quot;] = adata.X.copy()
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
/tmp/ipykernel_3037/1517723426.py:1: ImplicitModificationWarning: Setting element `.layers[&#39;counts&#39;]` of view, initializing view as actual.
  adata.layers[&#34;counts&#34;] = adata.X.copy()
</pre></div></div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[7]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>sc.pp.highly_variable_genes(
    adata,
    n_top_genes=2000,
    flavor=&quot;seurat_v3&quot;,
    subset=True,
    layer=&quot;counts&quot;
)
</pre></div>
</div>
</div>
</section>
<section id="Factor-analysis">
<h2>Factor analysis<a class="headerlink" href="#Factor-analysis" title="Permalink to this heading">#</a></h2>
<p>Next, we model the data using scPCA. One strategy is to designate the 0h condition as the reference and capture gene expression shifts after 1h and 4h as deviations from this baseline. This can be specified using the design formula <code class="docutils literal notranslate"><span class="pre">~</span> <span class="pre">stim</span></code>, similar to how one would define a model in a linear regression framework.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[8]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>m0 = scp.scPCA(
    adata,
    num_factors=10,
    layers_key=&#39;counts&#39;,
    loadings_formula=&#39;stim&#39;,
    intercept_formula=&#39;1&#39;,
    seed=3258035
)
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[9]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>m0.fit()
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
  0%|          | 0/5000 [00:00&lt;?, ?it/s]/g/huber/users/voehring/conda/scpca_dev/lib/python3.12/site-packages/scpca/models/scpca.py:68: UserWarning: The use of `x.T` on tensors of dimension other than 2 to reverse their shape is deprecated and it will throw an error in a future release. Consider `x.mT` to transpose batches of matrices or `x.permute(*torch.arange(x.ndim - 1, -1, -1))` to reverse the dimensions of a tensor. (Triggered internally at /pytorch/aten/src/ATen/native/TensorShape.cpp:4413.)
  α_rna = deterministic(&#34;α_rna&#34;, (1 / α_rna_inv).T)
Epoch: 4990 | lr: 1.00E-02 | ELBO: 24579048 | Δ_10: -454472.00 | Best: 23914962: 100%|██████████| 5000/5000 [02:21&lt;00:00, 35.28it/s]
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[10]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>m0.fit(lr=0.001, num_epochs=10000)
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
Epoch: 14990 | lr: 1.00E-03 | ELBO: 23639794 | Δ_10: -142348.00 | Best: 23275284: 100%|██████████| 10000/10000 [04:38&lt;00:00, 35.87it/s]
</pre></div></div>
</div>
<p>Let’s store the result in the <code class="docutils literal notranslate"><span class="pre">adata</span></code> object.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[11]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>m0.mean_to_anndata(model_key=&#39;m0&#39;, num_samples=10, num_split=512)
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
Predicting z for obs 48128-48265.: 100%|██████████| 95/95 [00:05&lt;00:00, 16.93it/s]
</pre></div></div>
</div>
<p>If we now take a close look at the adata object we will find that the factors and loadings were added to the <code class="docutils literal notranslate"><span class="pre">obsm</span></code> and <code class="docutils literal notranslate"><span class="pre">varm</span></code> fields respectively (<code class="docutils literal notranslate"><span class="pre">X_m0</span></code> and <code class="docutils literal notranslate"><span class="pre">W_m0</span></code>). Meta data related to the factorisation are stored under <code class="docutils literal notranslate"><span class="pre">adata.uns[{model_key}]</span></code>.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[12]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>adata
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[12]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
AnnData object with n_obs × n_vars = 48266 × 2000
    obs: &#39;Cell&#39;, &#39;stim&#39;, &#39;sample&#39;, &#39;maintype&#39;, &#39;celltype&#39;, &#39;subtype&#39;
    var: &#39;highly_variable&#39;, &#39;highly_variable_rank&#39;, &#39;means&#39;, &#39;variances&#39;, &#39;variances_norm&#39;
    uns: &#39;X_name&#39;, &#39;celltype_column&#39;, &#39;condition_column&#39;, &#39;main_assay&#39;, &#39;hvg&#39;, &#39;m0&#39;
    obsm: &#39;X_m0&#39;
    varm: &#39;W_m0&#39;
    layers: &#39;counts&#39;
</pre></div></div>
</div>
<p>Let’s check our factor representation by getting a visual overview using UMAP. We use the UMAP helper function of scpca to accomplish that.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[14]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>scp.tl.umap(adata, &#39;X_m0&#39;)
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[15]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>sc.pl.embedding(adata, basis=&#39;X_m0_umap&#39;, color=[&#39;maintype&#39;, &#39;stim&#39;], ncols=1)
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_hrvatin_26_0.png" src="../_images/notebooks_hrvatin_26_0.png" />
</div>
</div>
<p>We now try to assess the main factors of the decomposition.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[16]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>scp.pl.factor_embedding(adata, model_key=&#39;m0&#39;, factor=[0, 1, 2, 3], basis=&#39;X_m0_umap&#39;, ncols=2)
plt.tight_layout()
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_hrvatin_28_0.png" src="../_images/notebooks_hrvatin_28_0.png" />
</div>
</div>
<section id="Factor-0">
<h3>Factor 0<a class="headerlink" href="#Factor-0" title="Permalink to this heading">#</a></h3>
<p>We begin by examining factor 0 more closely. This factor seems to delineate variation between Excitatory/Interneurons and all other cell types.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[17]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>scp.pl.factor_strip(adata, model_key=&#39;m0&#39;, factor=[0], cluster_key=&#39;maintype&#39;, highlight=[&#39;Excitatory&#39;, &#39;Interneurons&#39;], s=2)
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[17]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
&lt;Axes: title={&#39;center&#39;: &#39;Factor 0&#39;}, xlabel=&#39;cluster&#39;, ylabel=&#39;Factor weight&#39;&gt;
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_hrvatin_30_1.png" src="../_images/notebooks_hrvatin_30_1.png" />
</div>
</div>
<p>Next we take a close look at the <em>negative</em> loading weights of factor 0 to learn more about the expression patterns in the myeloid cells of this dataset. We can extract those from the <code class="docutils literal notranslate"><span class="pre">adata</span></code> object using the <code class="docutils literal notranslate"><span class="pre">state_loadings</span></code> function. We use the argument <code class="docutils literal notranslate"><span class="pre">lowest=20</span></code> and <code class="docutils literal notranslate"><span class="pre">highest=0</span></code> to extract the genes with the largest negative weights.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[19]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>top_weights_fac0 = scp.tl.state_loadings(adata, &#39;m0&#39;, &#39;Intercept&#39;, 0, highest=0, lowest=20)
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[23]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>top_weights_fac0
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[23]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>gene</th>
      <th>magnitude</th>
      <th>weight</th>
      <th>type</th>
      <th>state</th>
      <th>factor</th>
      <th>index</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>Mal2</td>
      <td>10.400877</td>
      <td>-10.400877</td>
      <td>lowest</td>
      <td>0</td>
      <td>0</td>
      <td>1150</td>
    </tr>
    <tr>
      <th>1</th>
      <td>Slc6a7</td>
      <td>10.412316</td>
      <td>-10.412316</td>
      <td>lowest</td>
      <td>0</td>
      <td>0</td>
      <td>1699</td>
    </tr>
    <tr>
      <th>2</th>
      <td>Slc12a5</td>
      <td>10.412939</td>
      <td>-10.412939</td>
      <td>lowest</td>
      <td>0</td>
      <td>0</td>
      <td>1649</td>
    </tr>
    <tr>
      <th>3</th>
      <td>Trank1</td>
      <td>10.452961</td>
      <td>-10.452961</td>
      <td>lowest</td>
      <td>0</td>
      <td>0</td>
      <td>1889</td>
    </tr>
    <tr>
      <th>4</th>
      <td>Gabrg2</td>
      <td>10.481948</td>
      <td>-10.481948</td>
      <td>lowest</td>
      <td>0</td>
      <td>0</td>
      <td>729</td>
    </tr>
    <tr>
      <th>5</th>
      <td>Slc17a7</td>
      <td>10.504766</td>
      <td>-10.504766</td>
      <td>lowest</td>
      <td>0</td>
      <td>0</td>
      <td>1657</td>
    </tr>
    <tr>
      <th>6</th>
      <td>Tbr1</td>
      <td>10.507494</td>
      <td>-10.507494</td>
      <td>lowest</td>
      <td>0</td>
      <td>0</td>
      <td>1807</td>
    </tr>
    <tr>
      <th>7</th>
      <td>Adcy1</td>
      <td>10.556755</td>
      <td>-10.556755</td>
      <td>lowest</td>
      <td>0</td>
      <td>0</td>
      <td>116</td>
    </tr>
    <tr>
      <th>8</th>
      <td>Phf24</td>
      <td>10.618876</td>
      <td>-10.618876</td>
      <td>lowest</td>
      <td>0</td>
      <td>0</td>
      <td>1409</td>
    </tr>
    <tr>
      <th>9</th>
      <td>Rbfox3</td>
      <td>10.658419</td>
      <td>-10.658419</td>
      <td>lowest</td>
      <td>0</td>
      <td>0</td>
      <td>1531</td>
    </tr>
    <tr>
      <th>10</th>
      <td>Gpr88</td>
      <td>10.676279</td>
      <td>-10.676279</td>
      <td>lowest</td>
      <td>0</td>
      <td>0</td>
      <td>904</td>
    </tr>
    <tr>
      <th>11</th>
      <td>Kalrn</td>
      <td>10.729308</td>
      <td>-10.729308</td>
      <td>lowest</td>
      <td>0</td>
      <td>0</td>
      <td>1036</td>
    </tr>
    <tr>
      <th>12</th>
      <td>Myt1l</td>
      <td>10.771517</td>
      <td>-10.771517</td>
      <td>lowest</td>
      <td>0</td>
      <td>0</td>
      <td>1235</td>
    </tr>
    <tr>
      <th>13</th>
      <td>Vsnl1</td>
      <td>10.859935</td>
      <td>-10.859935</td>
      <td>lowest</td>
      <td>0</td>
      <td>0</td>
      <td>1954</td>
    </tr>
    <tr>
      <th>14</th>
      <td>Lamp5</td>
      <td>10.900677</td>
      <td>-10.900677</td>
      <td>lowest</td>
      <td>0</td>
      <td>0</td>
      <td>1088</td>
    </tr>
    <tr>
      <th>15</th>
      <td>Cck</td>
      <td>10.972982</td>
      <td>-10.972982</td>
      <td>lowest</td>
      <td>0</td>
      <td>0</td>
      <td>325</td>
    </tr>
    <tr>
      <th>16</th>
      <td>Sncb</td>
      <td>11.009183</td>
      <td>-11.009183</td>
      <td>lowest</td>
      <td>0</td>
      <td>0</td>
      <td>1720</td>
    </tr>
    <tr>
      <th>17</th>
      <td>Gabra1</td>
      <td>11.043131</td>
      <td>-11.043131</td>
      <td>lowest</td>
      <td>0</td>
      <td>0</td>
      <td>726</td>
    </tr>
    <tr>
      <th>18</th>
      <td>Nptx1</td>
      <td>11.110707</td>
      <td>-11.110707</td>
      <td>lowest</td>
      <td>0</td>
      <td>0</td>
      <td>1281</td>
    </tr>
    <tr>
      <th>19</th>
      <td>Sv2b</td>
      <td>11.851362</td>
      <td>-11.851362</td>
      <td>lowest</td>
      <td>0</td>
      <td>0</td>
      <td>1786</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
<p>Expectedly, we find that these these loadings indicate excitatory neurons.</p>
<table class="table">
<colgroup>
<col style="width: 33%" />
<col style="width: 67%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Gene</p></th>
<th class="head"><p>Annotation</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p><strong>Slc17a7 (VGLUT1)</strong></p></td>
<td><p>Marker of <strong>glutamatergic (excitatory) neurons</strong>, especially in cortex and hippocampus.</p></td>
</tr>
<tr class="row-odd"><td><p><strong>Tbr1</strong></p></td>
<td><p>Transcription factor enriched in <strong>layer 6 excitatory neurons</strong> of the cortex.</p></td>
</tr>
<tr class="row-even"><td><p><strong>Rbfox3 (NeuN)</strong></p></td>
<td><p>General <strong>neuronal marker</strong>, nuclear protein expressed in most mature neurons.</p></td>
</tr>
<tr class="row-odd"><td><p><strong>Gabra1</strong>, <strong>Gabrg2</strong></p></td>
<td><p>Encode GABA-A receptor subunits – expressed in both inhibitory and excitatory neurons (receptive to GABA).</p></td>
</tr>
<tr class="row-even"><td><p><strong>Cck</strong>, <strong>Lamp5</strong></p></td>
<td><p>Typically associated with <strong>interneurons</strong>, but also found in specific <strong>excitatory subtypes</strong>, particularly in superficial cortical layers.</p></td>
</tr>
<tr class="row-odd"><td><p><strong>Slc12a5 (KCC2)</strong></p></td>
<td><p>Potassium-chloride transporter critical for <strong>neuronal chloride homeostasis</strong> – enriched in mature <strong>excitatory neurons</strong>.</p></td>
</tr>
<tr class="row-even"><td><p><strong>Slc6a7</strong></p></td>
<td><p>High-affinity <strong>L-proline transporter</strong>, enriched in <strong>glutamatergic neurons</strong>, possibly in <strong>layer 5/6</strong>.</p></td>
</tr>
<tr class="row-odd"><td><p><strong>Mal2</strong>, <strong>Kalrn</strong>, <strong>Adcy1</strong>, <strong>Phf24</strong>, <strong>Trank1</strong>, <strong>Gpr88</strong>, <strong>Sncb</strong>, <strong>Myt1l</strong>, <strong>Vsnl1</strong>, <strong>Nptx1</strong>, <strong>Sv2b</strong></p></td>
<td><p>Involved in <strong>synaptic function</strong>, <strong>neuronal maturation</strong>, and <strong>plasticity</strong>. Most are enriched in <strong>excitatory neurons</strong> of the cortex.</p></td>
</tr>
</tbody>
</table>
<p>Next, we would like to understand how the factor 0 weights change across the condtitions. We provide the <code class="docutils literal notranslate"><span class="pre">scp.pl.loadings_scatter</span></code> function to visualise the loading weights across several condtions. This function requires us to specify the genes we want to visualise. To find interesting genes we focus on the genes with largest loading state differences between the 0h and 1h condition.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[33]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>loading_diff0 = scp.tl.state_diff(adata, model_key=&#39;m0&#39;, states=[&#39;Intercept&#39;, &#39;stim[T.1h]&#39;], factor=0, sign=-1)
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[42]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>scp.pl.loadings_scatter(adata, &#39;m0&#39;, [&#39;Intercept&#39;, &#39;stim[T.1h]&#39;, &#39;stim[T.4h]&#39;], 0, sign=-1, var_names=loading_diff0.gene.tolist(), show_labels=1)
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[42]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
&lt;Axes: title={&#39;center&#39;: &#39;Factor 0&#39;}, xlabel=&#39;State&#39;, ylabel=&#39;Loading weight&#39;&gt;
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_hrvatin_37_1.png" src="../_images/notebooks_hrvatin_37_1.png" />
</div>
</div>
<p>We find that this gene set is enriched for neuronal immediate early genes (IEGs) and activity-induced transcription factors.</p>
<table class="table">
<colgroup>
<col style="width: 12%" />
<col style="width: 88%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Gene</p></th>
<th class="head"><p>Function / Annotation</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p><strong>Fosb</strong>, <strong>Fosl2</strong></p></td>
<td><p>Core <strong>immediate early genes (IEGs)</strong>; rapidly induced by neuronal activity; components of the <strong>AP-1 transcription factor complex</strong>.</p></td>
</tr>
<tr class="row-odd"><td><p><strong>Egr2</strong>, <strong>Egr4</strong></p></td>
<td><p><strong>Early growth response</strong> transcription factors; induced by neuronal depolarization and activity.</p></td>
</tr>
<tr class="row-even"><td><p><strong>Nr4a2</strong></p></td>
<td><p><strong>Nuclear receptor</strong> involved in <strong>dopaminergic neuron function</strong>, plasticity, and survival.</p></td>
</tr>
<tr class="row-odd"><td><p><strong>Gadd45b</strong></p></td>
<td><p>Activity-dependent gene involved in <strong>DNA demethylation</strong> and <strong>neuroplasticity</strong>.</p></td>
</tr>
<tr class="row-even"><td><p><strong>Tnfaip6</strong></p></td>
<td><p>Induced by <strong>inflammation and neuronal stimulation</strong>; involved in <strong>extracellular matrix remodeling</strong>.</p></td>
</tr>
<tr class="row-odd"><td><p><strong>Bdnf</strong></p></td>
<td><p><strong>Brain-derived neurotrophic factor</strong>; crucial for <strong>synaptic plasticity, learning, and memory</strong>.</p></td>
</tr>
<tr class="row-even"><td><p><strong>Pim1</strong></p></td>
<td><p>Activity-induced <strong>serine/threonine kinase</strong>, regulates <strong>cell survival</strong> in neurons.</p></td>
</tr>
<tr class="row-odd"><td><p><strong>Nptx2</strong></p></td>
<td><p><strong>Neuronal pentraxin</strong>, secreted protein involved in <strong>synapse remodeling and homeostatic plasticity</strong>.</p></td>
</tr>
</tbody>
</table>
<p>Likewise we can try to elucdiate gene expression shifts between reference (baseline) and 4h condition.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[38]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>loading_diff0 = scp.tl.state_diff(adata, model_key=&#39;m0&#39;, states=[&#39;Intercept&#39;, &#39;stim[T.4h]&#39;], factor=0, sign=-1)
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[41]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>scp.pl.loadings_scatter(adata, &#39;m0&#39;, [&#39;Intercept&#39;, &#39;stim[T.1h]&#39;, &#39;stim[T.4h]&#39;], 0, sign=-1, var_names=loading_diff0.gene.tolist(), show_labels=2)
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[41]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
&lt;Axes: title={&#39;center&#39;: &#39;Factor 0&#39;}, xlabel=&#39;State&#39;, ylabel=&#39;Loading weight&#39;&gt;
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_hrvatin_41_1.png" src="../_images/notebooks_hrvatin_41_1.png" />
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>
</pre></div>
</div>
</div>
</section>
</section>
</section>


                </article>
              

              
              
              
              
                <footer class="prev-next-footer">
                  <!-- Previous / next buttons -->
<div class="prev-next-area">
    <a class="left-prev"
       href="angelidis.html"
       title="previous page">
      <i class="fa-solid fa-angle-left"></i>
      <div class="prev-next-info">
        <p class="prev-next-subtitle">previous</p>
        <p class="prev-next-title">Analysing the brain cells in light stimulated mice</p>
      </div>
    </a>
</div>
                </footer>
              
            </div>
            
            
              
                <div class="bd-sidebar-secondary bd-toc"><div class="sidebar-secondary-items sidebar-secondary__inner">

  <div class="sidebar-secondary-item">
  <div class="page-toc tocsection onthispage">
    <i class="fa-solid fa-list"></i> Contents
  </div>
  <nav class="bd-toc-nav page-toc">
    <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#Data-preprocessing">Data preprocessing</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#Factor-analysis">Factor analysis</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#Factor-0">Factor 0</a></li>
</ul>
</li>
</ul>
  </nav></div>

</div></div>
              
            
          </div>
          <footer class="bd-footer-content">
            
<div class="bd-footer-content__inner container">
  
  <div class="footer-item">
    
<p class="component-author">
By Harald Vohringer
</p>

  </div>
  
  <div class="footer-item">
    
  <p class="copyright">
    
      © Copyright 2023, Harald Vohringer.
      <br/>
    
  </p>

  </div>
  
  <div class="footer-item">
    
  </div>
  
  <div class="footer-item">
    
  </div>
  
</div>
          </footer>
        

      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="../_static/scripts/bootstrap.js?digest=927b94d3fcb96560df09"></script>
<script src="../_static/scripts/pydata-sphinx-theme.js?digest=927b94d3fcb96560df09"></script>

  <footer class="bd-footer">
  </footer>
  </body>
</html>